src/derivative_utils.jl

```diff
@@ -372,78 +387,18 @@ function calc_W!(W, integrator, cache, dtgamma, repeat_step, W_transform=false)
     return nothing
   end
 
-  # fast pass
-  # we only want to factorize the linear operator once
-  new_jac = true
-  new_W = true
-  if (f isa ODEFunction && islinear(f.f)) || (integrator.alg isa SplitAlgorithms && f isa SplitFunction && islinear(f.f1.f))
-    new_jac = false
-    @goto J2W # Jump to W calculation directly, because we already have J
-  end
-
   # check if we need to update J or W
-  W_dt = isnewton ? cache.nlsolver.cache.W_dt : dt # TODO: RosW
-  new_jac = isnewton ? do_newJ(integrator, alg, cache, repeat_step) : true
-  new_W = isnewton ? do_newW(integrator, cache.nlsolver, new_jac, W_dt) : true
+  W_dt = nlsolver === nothing ? dt : nlsolver.cache.W_dt # TODO: RosW
```

Comment:
I get `nlsolver === nothing` is a check for Rosenbrock, but the code might read better if we somehow formalize this notion to semi-implicit nlsolver. FIRK is another special case. We should follow up with @kanav99's there in a future PR. CC @devmotion